### ISSUE TITLE
Adding kops cluster to federated cluster

### ISSUE THREAD

**Issue by treacher** [kubernetes/kops#2626]
Hi there, I recently looked at adding a kops cluster to a federated cluster and I ran into a problem due to the nature around having to name clusters with a FQDN. Please see the example below:
Running:
`kubefed join cluster.foo.bar.com --host-cluster-context=cluster2.foo.bar.com`
Results in:
`The Cluster "cluster.foo.bar.com" is invalid: metadata.name: Invalid value: "cluster.foo.bar.com": a DNS-1035 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')`

Is there a way to not use a FQDN for the `metadata.name` of the cluster in kops? Or have I missed something completely.

Thanks

**Comment by chrislovecnm** [comment]
That actually looks like s bug with federation ;( Why can't we name our clusters with dns names?

**Comment by treacher** [comment]
I don't believe it's a bug given it's quite explicit in the documentation: https://kubernetes.io/docs/tasks/federation/set-up-cluster-federation-kubefed/#naming-rules-and-customization

I suspect when clusters are joined to federated clusters under the hood there are some form of DNS records created for referential purposes and if you have an FQDN as the cluster name it will cause issues and that's why they document they need to follow subdomain naming conventions rather than to be an FQDN.

**Comment by chrislovecnm** [comment]
Are you doing the API server fqdn? I do not have much fed experience at all, can you provide some example of what it expencts? What would work? ELI5

The API servers have to resolve, so I am confused how that cannot be a fqdn.

**Comment by treacher** [comment]
Sure, I'll try to explain what I understand and please correct me if I'm wrong:

* Kops dictates that the name of the cluster has to be an FQDN. [reference](https://github.com/kubernetes/kops/blob/master/pkg/apis/kops/validation/legacy.go#L69)
* Federated Kubernetes requires the cluster name follow [RFC-1035](https://www.ietf.org/rfc/rfc1035.txt) which is a subdomain not an FQDN. [reference](https://kubernetes.io/docs/tasks/federation/set-up-cluster-federation-kubefed/#naming-rules-and-customization)

There is a mismatch here in regards to the naming conventions defined between KOPS and what Federated Clusters require for cluster names.

From what I can see, the cluster name and the api servers FQDN don't need to be coupled but within kops they currently are.

Given you can specify the FQDN for the API servers separately in the kops cluster config, why does the cluster name need to be a FQDN?

I believe the solution to get it working with federated clusters is to remove the need for the cluster name to be an FQDN within kops.

**Comment by jvalencia** [comment]
I just tried to set up with kubefed and kops and failed as well, I'm sorting through issues-- this is one. I imagine this needs a fair bit work to triage what supporting federation looks like.

**Comment by qqshfox** [comment]
`--cluster-context` solve this problem for me.

```
kubefed join aws-cn-north-1-test2 --host-cluster-context=aws-cn-north-1.test1.k8s.local --cluster-context=aws-cn-north-1.test2.k8s.local
```
